### DISCUSSION TITLE
hi, my 'row.pin('top')' isn't working

### DISCUSSION THREAD

**Discussion by sparrowscout** [TanStack/table#5491] [Q&A]
1. table options : `enableRowPinning: true` , `keepPinnedRows: true` , `onRowPinningChange: setRowPinning`, `state:{rowPinning}`
2. column options  : `enablePinning: true`
3. in jsx : <br/>
``` 
...
<tbody>
{table.getRowModel().rows.map((row) => {
            if (
              !row.getIsPinned() &&
              row.getValue('role') === memberRoleInfo.OWNER
            ) {
              // I checked here ✅
              row.pin('top')
            }
return ( <tr> ...
```
4. and console.log(rowPinning) : <br/>
```
// I don't know why is it different from tanstack example ...
// examples like -> `{ "top": [ "0", "0.0", "0.0.0", "0.0.1", "0.1", "0.1.0", "0.1.1" ], "bottom": [] }`
{
    top: ["0"],
    bottom: [],
  }
```
5. and it is not working ... could you help me? am I missing something??
6. I am using Next.js 14

**Comment by zahidiqbalnbs** [comment]
@sparrowscout any luck with this?

**Reply by KevinVandy** [reply]
Share a full sandbox?

**Reply by zahidiqbalnbs** [reply]
Hi @KevinVandy  I hope you are doing well. Thank you for implementing the great feature of row pinning.

Here is the [sandbox](https://codesandbox.io/p/sandbox/react-table-default-row-pinning-khy9cr).

You will see I am setting the few entries as a pinned rows in a way mentioned in sandbox in default state. It is getting pinned but pagination is messing with use of this feature. Total 26 records, first page showing 13, second one 7 and third one 5.

Please let me know if I implemented in correct way or is there any other way of setting the default pinned rows when mounting table?

**Reply by KevinVandy** [reply]
do these work as expected for you? https://www.material-react-table.dev/?path=/story/features-row-pinning-examples--row-pinning-sticky-default-enabled

**Reply by zahidiqbalnbs** [reply]
No. No example showing how to make few rows pinning enabled as a default state when Table is mounting. I do not want to use Button to pin that row which is shown in example on tanstack website. Also when sorting is getting applied whole pagination is getting disturbed as you can see in the sandbox with rows pinning feature.

**Reply by KevinVandy** [reply]
So just initialState?

**Reply by zahidiqbalnbs** [reply]
Yes, have you seen the sandbox that I have shared and the pagination issue which is happening with row pinning functionality?

**Reply by KevinVandy** [reply]
I'm not sure I see any unexpected behavior after getting rid of `table.getState().pagination.pageIndex <= 0 &&` on line 106. If you call `table.getCenterRows`, it will filter out pinned rows, and it looks like it's doing that correctly.

You can change `table.getCenterRows` back to `table.getRowModel().rows` if you want to keep the pinned rows in the main body too.

**Reply by zahidiqbalnbs** [reply]
@KevinVandy I want to pin rows on top of the entire table, lets say if there are 50 records and I want to pin first 2 records so that should only stay on top (means on first page), when going to second page those 2 pin records should not show. And should also predict the same behavior if I am applied sorting. Sorting should not work on those 2 records but for other 48 records it should work. Is this doable using pinned rows feature?

**Reply by KevinVandy** [reply]
You can probably write a custom paginated row model for that. Just fork the stock one, add some logic to filter out pinned rows, then let the normal pagination proceed.

Actually you could probably put the logic in the core row model instead

**Reply by zahidiqbalnbs** [reply]
@KevinVandy can you kindly create the sandbox for me if you have some time. would be highly appreciated. Thanks in advance!

**Reply by KevinVandy** [reply]
IDK when, but it might be just a couple lines of code with a filter. The row model docs might be helpful
